### PR TITLE
Feature Login

### PR DESCRIPTION
**Pull request by dangcongkhaithu**
Register
Login
Change Language
Add to Watchlist

### CODE REVIEW

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -14,8 +14,27 @@
     <string name="play_trailer">Play Trailer</string>
     <string name="user_score">User Score</string>
     <string name="overview">Overview</string>
-    <string name="default_hint_search">Search TMDb</string>
+    <string name="defaultHintSearch">Search TMDb</string>
```
Comment (vuphu-jc) [inline app/src/main/res/values/strings.xml]: keep old

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -14,8 +14,27 @@
...
+    <string name="login">Login</string>
+    <string name="cancel">Cancel</string>
+    <string name="confirm_password">confirm password</string>
+    <string name="don_t_have_an_account">Don\'t have an account,</string>
+    <string name="full_name">full name</string>
+    <string name="email">email</string>
+    <string name="get_personalized_recommendations_and_keep_track_of_what_to_watch">Get personalized recommendations and keep track of what to watch</string>
+    <string name="add_list">Add to watch list</string>
+    <string name="your_watch_list">Your watchlist</string>
+    <string name="_0">0</string>
```
Comment (vuphu-jc) [inline app/src/main/res/values/strings.xml]:
```suggestion
    <string name="zero">0</string>
```

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/customview/VintageClockView.kt [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        drawBorder(canvas, radiusCircle2)
+        drawBorder(canvas, radiusCircle3)
+        drawBorder(canvas, radiusCircle4)
+        drawPointCircle(canvas, radiusWidthParent, radiusHeightParent, radiusPointCenter)
+        drawTimeClock(canvas)
+        setTimeForClock(canvas)
+
+        Handler().postDelayed({
+            invalidate()
+        }, 1000)
```
Comment (chuongdevJC) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/customview/VintageClockView.kt]: should be define a variable to keep instant to cancel later

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/RegisterActivity.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_register)
+
+        buttonRegister.setOnClickListener(View.OnClickListener {
+            val user: User = User(
+                editFullName.text.toString(),
+                editEmail.text.toString()
+            )
+            val password : String =   editPassword.text.toString()
```
Comment (vuphu-jc) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/RegisterActivity.kt]:
```suggestion
            val password : String = editPassword.text.toString()
```

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/interfacePre/IUserPresenter.kt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package com.khaithu.a3ctech_android_final_project.presenter.interfacePre
+
+interface IUserPresenter {
+    fun getTitle()
+
+    fun createWatchlist()
+
+    fun removeItemWatchlist(id : Int)
```
Comment (vuphu-jc) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/interfacePre/IUserPresenter.kt]:
```suggestion
    fun removeWatchlist(id : Int)
```

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+
+    override fun createWatchlist() {
+        firebaseAuth = FirebaseAuth.getInstance()
+        rootNode = FirebaseDatabase.getInstance()
+        reference = rootNode.reference
+
+        firebaseAuth.currentUser?.uid?.let {
+            reference.child("User").child(it).child("Watchlist")
```
Comment (vuphu-jc) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt]: you can define some firebase keys such as "User", "Watchlist" at new class `FirebaseConstant`

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/GetMovieDetailPresenter.kt [hunk 7]
```diff
@@ -50,6 +59,25 @@ class GetMovieDetailPresenter(var iMovieDetailView: IMovieDetailView) : BasePres
             })
     }
 
+    override fun addToWatchlist(movieDetail: MovieDetail) {
+        firebaseAuth = FirebaseAuth.getInstance()
+        rootNode = FirebaseDatabase.getInstance()
+        reference = rootNode.getReference("User")
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/GetMovieDetailPresenter.kt]: define these nodes

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    private lateinit var reference: DatabaseReference
+    private lateinit var firebaseAuth: FirebaseAuth
+
+    override fun getTitle() {
+        firebaseAuth = FirebaseAuth.getInstance()
+        rootNode = FirebaseDatabase.getInstance()
+        reference = rootNode.reference
+
+        firebaseAuth.currentUser?.uid?.let {
+            reference.child("User").child(it).child("fullName").get()
```
Comment (vuphu-jc) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt]: define "User" and "fullName" at constant

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/GetMovieDetailPresenter.kt [hunk 9]
```diff
@@ -50,6 +59,25 @@ class GetMovieDetailPresenter(var iMovieDetailView: IMovieDetailView) : BasePres
...
+        val watchlistItem: WatchlistItem = WatchlistItem(
+            movieDetail.id,
+            movieDetail.title,
+            movieDetail.releaseDate,
+            movieDetail.posterPath,
+            movieDetail.voteAverage
+        )
+
+        firebaseAuth.currentUser?.uid?.let {
+            reference.child(it).child("Watchlist").child(movieDetail.id.toString())
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/GetMovieDetailPresenter.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt (hunk 8 above)
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt (hunk 6 above)
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+
+    override fun removeItemWatchlist(id : Int) {
+        firebaseAuth = FirebaseAuth.getInstance()
+        rootNode = FirebaseDatabase.getInstance()
+        reference = rootNode.reference
+
+        firebaseAuth.currentUser?.uid?.let {
+            reference.child("User").child(it).child("Watchlist")
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/interfaceView/IRegisterActivity.kt [hunk 11]
```diff
@@ -0,0 +1,6 @@
+package com.khaithu.a3ctech_android_final_project.view.interfaceView
+
+interface IRegisterActivity {
+    fun registerSuccess()
+    fun registerFail()
+}
```
Comment (vuphu-jc) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/interfaceView/IRegisterActivity.kt]:
```suggestion
interface IRegisterView {
    fun onRegisterSuccess()
    fun onRegisterFail()
}
```

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/sharedprefences/MyApplication.kt [hunk 12]
```diff
@@ -1,10 +1,14 @@
...
 
 import android.app.Application
+import com.khaithu.a3ctech_android_final_project.model.enum.LanguageEnum
 
-class MyApplication: Application() {
+class MyApplication : Application() {
     override fun onCreate() {
         super.onCreate()
         DataLocalManager.init(applicationContext)
+        if (DataLocalManager.getLanguage() == "") {
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/sharedprefences/MyApplication.kt]: use isBlank instead pf

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/MovieDetailActivity.kt [hunk 13]
```diff
@@ -67,7 +84,7 @@ class MovieDetailActivity : AppCompatActivity(), IMovieDetailView {
 
     private fun handleDisplayTitle(movieDetail: MovieDetail) {
         if (movieDetail.releaseDate.isEmpty() || movieDetail.title.isEmpty()) {
-            titleDetailMovie.gone()
+            titleDetailMovie.visibility = View.GONE
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/MovieDetailActivity.kt]: why you change it?

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/RegisterActivity.kt (hunk 4 above)
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/RegisterActivity.kt]: format

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/dialog/LoginDialog.kt [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+class LoginDialog : DialogFragment(), ILoginDialog {
+
+    private var mPresenter: LoginPresenter = LoginPresenter(this)
+    private var mProgressBar: ProgressBar? = null
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        val view: View? = activity?.layoutInflater?.inflate(R.layout.login_dialog, null)
+        val builder: AlertDialog.Builder? = activity?.let { AlertDialog.Builder(it) }
+        builder?.setTitle("Login")
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/dialog/LoginDialog.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/RegisterActivity.kt [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+            mPresenter.register(user, password)
+        })
+
+        buttonCancel.setOnClickListener(View.OnClickListener {
+            finish()
+        })
+    }
+
+    override fun registerSuccess() {
+        Toast.makeText(this, "Register success", Toast.LENGTH_LONG).show()
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/RegisterActivity.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/RegisterActivity.kt [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+        finish()
+    }
+
+    override fun registerFail() {
+        Toast.makeText(this, "Register fail", Toast.LENGTH_LONG).show()
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/RegisterActivity.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/sharedprefences/DataLocalManager.kt [hunk 17]
```diff
@@ -19,5 +22,21 @@ class DataLocalManager {
...
+            return newInstance().mySharedPreference.getStringValue(LANGUAGE)
+        }
+
+        fun setStatusLogin(status: Boolean) {
+            newInstance().mySharedPreference.putBooleanValue(STATUS_LOGIN, status)
+        }
+
+        fun getStatusLogin() : Boolean {
+            return newInstance().mySharedPreference.getBooleanValue(STATUS_LOGIN)
+        }
```
Comment (vuphu-jc) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/sharedprefences/DataLocalManager.kt]:
```suggestion
        fun setLoginStatus(status: Boolean) {
            newInstance().mySharedPreference.putBooleanValue(STATUS_LOGIN, status)
        }

        fun getLoginStatus() : Boolean {
            return newInstance().mySharedPreference.getBooleanValue(STATUS_LOGIN)
        }
```

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/dialog/LoginDialog.kt [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+        return alertDialog!!
+    }
+
+    override fun loginSuccess() {
+        DataLocalManager.setStatusLogin(true)
+        dismiss()
+    }
+
+    override fun loginFail() {
+        Toast.makeText(context, "Login fail", Toast.LENGTH_LONG).show()
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/dialog/LoginDialog.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/sharedprefences/DataLocalManager.kt [hunk 19]
```diff
@@ -1,15 +1,18 @@
...
 import android.content.Context
+import com.khaithu.a3ctech_android_final_project.model.enum.LanguageEnum
 
 class DataLocalManager {
     private lateinit var mySharedPreference: MySharedPreference
 
     companion object {
         private lateinit var instance: DataLocalManager
+        private const val LANGUAGE = "LANGUAGE"
+        private const val STATUS_LOGIN = "STATUS_LOGIN"
```
Comment (vuphu-jc) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/sharedprefences/DataLocalManager.kt]:
```suggestion
        private const val LANGUAGE_KEY = "LANGUAGE_KEY"
        private const val LOGIN_STATUS_KEY = "LOGIN_STATUS_KEY"
```
or define at new class SharedPrefsKey

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/navigator/MovieDetailNavigator.kt [hunk 20]
```diff
@@ -14,4 +15,11 @@ class MovieDetailNavigator(var mMovieDetailActivity: MovieDetailActivity) : IMov
         ContextCompat.startActivity(mMovieDetailActivity, intent, null)
         mMovieDetailActivity.overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
     }
+
+    override fun goToLoginDialog() {
+        val mLoginDialog = LoginDialog()
+        mLoginDialog.show(mMovieDetailActivity.supportFragmentManager, "")
+    }
+
+
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/navigator/MovieDetailNavigator.kt]: remove

File: app/src/main/res/layout/activity_greet.xml [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:background="@color/black_blue"
+    tools:context=".view.GreetActivity">
+
+    <com.khaithu.a3ctech_android_final_project.customview.VintageClockView
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"/>
+
+
```
Comment (AnhJobChat) [inline app/src/main/res/layout/activity_greet.xml:14]: .

File: app/src/main/res/layout/fragment_search_view.xml [hunk 22]
```diff
@@ -6,59 +6,64 @@
...
 
-        </RelativeLayout>
+                    <androidx.recyclerview.widget.RecyclerView
+                        android:id="@+id/pageRecycler"
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:layout_marginLeft="@dimen/dp_80"
+                        android:layout_marginRight="@dimen/dp_80"/>
+
+
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_search_view.xml]: .

File: app/src/main/res/layout/fragment_user_logged_out_view.xml [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+       android:layout_width="match_parent"
+       android:layout_height="wrap_content"
+       android:orientation="vertical">
+
+      <androidx.appcompat.widget.Toolbar
+          android:id="@+id/toolBar"
+          android:layout_width="match_parent"
+          android:layout_height="@dimen/dp_60"
+          android:background="@color/black_blue"
+          app:title="Sign In"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/fragment_user_logged_out_view.xml]: .

File: app/src/main/res/layout/item_pagination_number.xml [hunk 24]
```diff
@@ -8,5 +8,6 @@
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
         android:textColor="@color/black"
+        android:text="5"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/item_pagination_number.xml]: .

File: app/src/main/res/layout/item_watchlist.xml [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+                tools:ignore="RelativeOverlap,SpUsage" />
+
+            <TextView
+                android:id="@+id/textRemove"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_alignParentRight="true"
+                app:drawableLeftCompat="@drawable/ic_remove" />
+
+
```
Comment (AnhJobChat) [inline app/src/main/res/layout/item_watchlist.xml]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/WatchlistRecyclerAdapter.kt [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+                mIWatchListEvent.handleEventRemove(movie.id)
+            })
+        }
+    }
+}
+interface IWatchListEvent {
+    fun handleEvent(id :Int?)
+
+    fun handleEventRemove(id : Int?)
+}
```
Comment (vuphu-jc) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/WatchlistRecyclerAdapter.kt]: You should define clearly name of method. Easy to maintain later
My suggestion: 
- Name of callback class should be ended with suffix `Listener`, example `MovieWatchlistListener`
- Name of callback method should be started with prefix `on`
`fun onMovieWatchlistClicked(id: Int?)`
`fun onMovieWatchlistRemoved(id: Int?)`
*pls format code this file*

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+                .addListenerForSingleValueEvent(object : ValueEventListener {
+                    override fun onDataChange(snapshot: DataSnapshot) {
+                        snapshot.children.forEach { task ->
+                            mIUserLoggedInView.updateWatchlist(task.getValue(WatchlistItem::class.java)!!)
```
Comment (vuphu-jc) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/UserPresenter.kt]: Don't force wrap `!!`. Using `?.let` instead of

**Review by vuphu-jc — APPROVED** [review]

**Review by chuongdevJC — APPROVED** [review]

**Review by AnhJobChat — APPROVED** [review]

**Pull request merged**